File: Solves/leetcode/q_102.cxx
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(root==nullptr)
                return {};
        vector<vector<int>> ans;
        queue<TreeNode*> curr;
        curr.push(root);
        while(!curr.empty())
        {
            int depth=curr.size();
            vector<int> level;
            for(int i=0;i<depth;i++)
            {
                TreeNode* node=curr.front();
                curr.pop();
                level.push_back(node->val);
                if(node->left!=nullptr)
                    curr.push(node->left);
                if(node->right!=nullptr)
                    curr.push(node->right);
            }
            ans.push_back(level);
        }
        return ans;
    }
};

File: Solves/leetcode/q_102.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        final List<List<Integer>> res = new ArrayList<>();
        if(root == null) {
            return res;
        }
        final var queue = new ArrayDeque<TreeNode>();
        queue.offerFirst(root);
        while(!queue.isEmpty()) {
            final var list = new ArrayList<Integer>();
            final var depth = queue.size();
            for(int i = 0; i < depth; i++) {
                final var current = queue.pollLast();
                list.add(current.val);
                if(current.left != null) {
                    queue.offerFirst(current.left);
                }
                if(current.right != null) {
                    queue.offerFirst(current.right);
                }
            }
            res.add(list);
        }
        return res;
    }
}
